org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java

```diff
@@ -302,30 +304,53 @@ public void onAttributeName(boolean generateValue, ICompletionRequest request, I
 	private void fillAttributesWithCMAttributeDeclarations(DOMElement parentElement, Range fullRange,
 			CMElementDeclaration elementDeclaration, boolean canSupportSnippet, boolean generateValue,
 			ICompletionRequest request, ICompletionResponse response) {
-
+		if (parentElement == null) {
+			return;
+		}
 		Collection<CMAttributeDeclaration> attributes = elementDeclaration.getAttributes();
 		if (attributes == null) {
 			return;
 		}
 		for (CMAttributeDeclaration attributeDeclaration : attributes) {
-			if (parentElement != null) {
-				String prefix = parentElement.getPrefix(attributeDeclaration.getNamespace());
-				String attrName = attributeDeclaration.getName(prefix);
-				if (!parentElement.hasAttribute(attrName)) {
-					CompletionItem item = new AttributeCompletionItem(attrName, canSupportSnippet, fullRange,
-							generateValue,
-							attributeDeclaration.getDefaultValue(), attributeDeclaration.getEnumerationValues(),
-							request.getSharedSettings());
-					if (request.isResolveDocumentationSupported()) {
-						addResolveData(request, item, AttributeNameCompletionResolver.PARTICIPANT_ID);
-					} else {
-						MarkupContent documentation = XMLGenerator.createMarkupContent(attributeDeclaration,
-								elementDeclaration,
-								request);
-						item.setDocumentation(documentation);
-					}
-					response.addCompletionAttribute(item);
+			String prefix = null;
+			boolean declareNamespace = false;
+			String namespaceURI = attributeDeclaration.getNamespace();
+			if (!StringUtils.isEmpty(namespaceURI)) {
+				prefix = parentElement.getPrefix(namespaceURI);
+				if (prefix == null) {
+					prefix = attributeDeclaration.getPrefix();
+					declareNamespace = true;
+				}
+			}
+			String attrName = attributeDeclaration.getName(prefix);
+			if (!parentElement.hasAttribute(attrName)) {
+				CompletionItem item = new AttributeCompletionItem(attrName, canSupportSnippet, fullRange,
+						generateValue,
+						attributeDeclaration.getDefaultValue(), attributeDeclaration.getEnumerationValues(),
+						request.getSharedSettings());
+				if (declareNamespace) {
+					DOMDocument document = parentElement.getOwnerDocument();
+					int offset = document.getDocumentElement().getOffsetBeforeCloseOfStartTag();
+					Range range = XMLPositionUtility.createRange(offset, offset, document);
+					StringBuilder insertText = new StringBuilder();
+					insertText.append("xmlns:");
+					insertText.append(prefix);
+					insertText.append("=\"");
+					insertText.append(namespaceURI);
+					insertText.append("\"");
+					TextEdit edit = new TextEdit(range, insertText.toString());
+					item.setAdditionalTextEdits(new ArrayList<>());
+					item.getAdditionalTextEdits().add(edit);

```

Comment:
item.setAdditionalTextEdits(Collections.singletonList(edit)); is more efficient

Reply:
fixed